### PR TITLE
Ccp 128

### PR DESCRIPTION
**Pull request by mkarthik90**
[CCP-128] Changes for indexing through lucene

### CODE REVIEW

File: Conceptpower+Spring/pom.xml [hunk 1]
```diff
@@ -23,8 +23,9 @@
 		<email.server.host>http://example.net/</email.server.host>
 		<admin.username>admin</admin.username>
 		<admin.password>admin</admin.password>
-		<wordnet.path>/usr/local/WordNet-3.0</wordnet.path>
+		<wordnet.path>C:/Program Files (x86)/WordNet/2.1</wordnet.path>
```
Comment (jdamerow) [inline Conceptpower+Spring/pom.xml]: are you wokring ona windows machine?

File: Conceptpower+Spring/src/main/resources/lucene.properties [hunk 2]
```diff
@@ -0,0 +1 @@
+lucenePath=${lucene.path}
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/resources/lucene.properties]: put this into config.properties. no need for a second config file.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+                try {
+                    Object contentOfField = field.get(entry);
+                    if (contentOfField != null) {
+                        doc.add(new StringField(searchFieldAnnotation.lucenefieldName(),
+                                String.valueOf(contentOfField).replace(" ", ""), Field.Store.YES));
+                    }
+
+                } catch (IllegalArgumentException ie) {
+                    throw new IllegalArgumentException(ie);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: either create a new exception with a custom exception class or throw the existing one, but this doesn't make sense.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+            }
+        }
+
+        doc.add(new StoredField(LuceneFieldNames.TYPE_ID, entry.getTypeId() != null ? entry.getTypeId() : ""));
+        doc.add(new StoredField(LuceneFieldNames.CREATOR, entry.getCreatorId() != null ? entry.getCreatorId() : ""));
+        doc.add(new StoredField(LuceneFieldNames.MODIFIED, entry.getModified() != null ? entry.getModified() : ""));
+        Calendar cal = Calendar.getInstance();
+        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSSZ");
+        doc.add(new StoredField(LuceneFieldNames.MODIFIED_TIME, formatter.format(cal.getTime())));
+        doc.add(new StoredField(LuceneFieldNames.ID, entry.getId() != null ? entry.getId() : ""));
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: why do you do line 122-128? this should be covered by the above generic code.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+        for (java.lang.reflect.Field field : fields) {
+            LuceneField luceneFieldAnnotation = field.getAnnotation(LuceneField.class);
+            field.setAccessible(true);
+            try {
+                if (luceneFieldAnnotation != null && d.get(luceneFieldAnnotation.lucenefieldName()) != null)
+                    field.set(con, d.get(luceneFieldAnnotation.lucenefieldName()));
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException(e);
+            } catch (IllegalAccessException e) {
+                throw new IllegalAccessException();
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: same here

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        } catch (IOException ex) {
+            throw new LuceneException("Cannot insert concept in lucene. Please retry", ex);
+        }
+
+    }
+
+    private ConceptEntry getConceptFromDocument(Document d) throws IllegalAccessException {
+
+        ConceptEntry con = new ConceptEntry();
+        java.lang.reflect.Field[] fields = con.getClass().getDeclaredFields();
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: you don't need a to create an object for this. just do:
```
ConceptEntry.class.getDeclaredFields();
```

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        con.setId(d.get("id"));
+        return con;
+    }
+
+    @Override
+    public void deleteIndexes() throws LuceneException {
+        try {
+            Query q = new QueryParser("conceptType", whiteSpaceAnalyzer).parse("*:*");
+            writer.deleteDocuments(q);
+            writer.commit();
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: This method deletes all indexes, right? You can probably do that more performant by opening the index in create mode:
```
IndexWriterConfig configWhiteSpace = new IndexWriterConfig(whiteSpaceAnalyzer);
configWhiteSpace.setOpenMode(OpenMode.CREATE);
```
(CREATE - creates a new index or overwrites an existing one)

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+    @Override
+    public void indexConcepts() throws LuceneException {
+
+        try {
+            String wnhome = configuration.getWordnetPath();
+            String path = wnhome + File.separator + configuration.getDictFolder();
+
+            URL url = null;
+
+            url = new URL("file", null, path);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: merge line with 214

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+            iterator = dict.getIndexWordIterator(POS.ADVERB);
+            createDocuments(iterator, dict, writer);
+
+            iterator = dict.getIndexWordIterator(POS.ADJECTIVE);
+            createDocuments(iterator, dict, writer);
+
+            iterator = dict.getIndexWordIterator(POS.VERB);
+            createDocuments(iterator, dict, writer);
+
+            writer.commit();
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: I'd commit the writer after every call to createdocuments to reduce memory usage

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+                for (IWord syn : synonyms) {
+                    if (!syn.getID().equals(word.getID()))
+                        sb.append(
+                                syn.getID().toString() + edu.asu.conceptpower.servlet.core.Constants.SYNONYM_SEPARATOR);
+                }
+                doc.add(new StringField(LuceneFieldNames.SYNONYMID, sb.toString(), Field.Store.YES));
+                // Adding this new data to delete only wordnet concepts while
+                // adding all wordnet concepts from jwi.
+                doc.add(new StringField(LuceneFieldNames.CONCEPTTYPE, "wordnetconcept", Field.Store.YES));
+                writer.addDocument(doc);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: I wouldn't throw the exception, handle it. otherwise if Lucene can't write one of the words, it'll give up on all of them. I'd return a result object that says how many words couldn't be indexed.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+
+            iterator = dict.getIndexWordIterator(POS.ADJECTIVE);
+            createDocuments(iterator, dict, writer);
+
+            iterator = dict.getIndexWordIterator(POS.VERB);
+            createDocuments(iterator, dict, writer);
+
+            writer.commit();
+
+        } catch (IOException ex) {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: the catch should be where the exception is actually thrown so that at the least the words that are indexable are stored.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+                    firstEntry++;
+                    queryString.append(luceneFieldAnnotation.lucenefieldName() + ":");
+                    queryString.append(searchString);
+                }
+            }
+        }
+
+        List<ConceptEntry> concepts = new ArrayList<ConceptEntry>();
+        int hitsPerPage = 10;
+        IndexSearcher searcher = null;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: This should be a class variable. From the documentation:
"For performance reasons, if your index is unchanging, you should share a single IndexSearcher instance across multiple searches instead of creating a new one per-search. If your index has changed and you wish to see the changes reflected in searching, you should use DirectoryReader.openIfChanged(DirectoryReader) to obtain a new reader and then create a new IndexSearcher from that."

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+                        queryString.append(" " + operator + " ");
+                    firstEntry++;
+                    queryString.append(luceneFieldAnnotation.lucenefieldName() + ":");
+                    queryString.append(searchString);
+                }
+            }
+        }
+
+        List<ConceptEntry> concepts = new ArrayList<ConceptEntry>();
+        int hitsPerPage = 10;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: this value should be configurable

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+import edu.mit.jwi.IDictionary;
+import edu.mit.jwi.item.IIndexWord;
+import edu.mit.jwi.item.ISynset;
+import edu.mit.jwi.item.IWord;
+import edu.mit.jwi.item.IWordID;
+import edu.mit.jwi.item.POS;
+
+@Component
+@PropertySource("classpath:lucene.properties")
+public class LuceneUtility implements ILuceneUtility {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: this class needs javadoc

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+     * @see
+     * edu.asu.conceptpower.core.IConceptManager#getConceptListEntriesForWord(
+     * java.lang.String, java.lang.String)
+     */
+    @Override
+    public ConceptEntry[] getConceptListEntriesForWord(String word, String pos, String conceptType)
+            throws LuceneException, IllegalAccessException {
+        if (pos == null)
+            return null;
+        word = word.replace(" ", "");
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java]: why do you do this?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+        Map<String,String> fieldMap = new HashMap<String,String>();
+        fieldMap.put(SearchFieldNames.WORD, word);
+        fieldMap.put(SearchFieldNames.POS, pos);
+        fieldMap.put(LuceneFieldNames.CONCEPTTYPE, conceptType);
+        return luceneUtility.queryIndex(fieldMap, null);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java]: why do you use lucenefieldnames here? and why did you add a concepttype? If you want to be able to search all wordnet concepts (from your calls it looks liek it), you can use the conceptlist property.

Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java]: instead of calling the lucene utility here, I'd call the search manager

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java [hunk 17]
```diff
@@ -0,0 +1,428 @@
...
+     * 
+     * @see
+     * edu.asu.conceptpower.core.IConceptManager#getConceptListEntriesForWord(
+     * java.lang.String)
+     */
+    @Override
+    public ConceptEntry[] getConceptListEntriesForWord(String word) throws LuceneException, IllegalAccessException {
+        Map<String,String> fieldMap = new HashMap<String,String>();
+        fieldMap.put(SearchFieldNames.WORD, word);
+        return luceneUtility.queryIndex(fieldMap, null);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java]: instead of calling the lucene utility here, I'd call the search manager

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java [hunk 18]
```diff
@@ -0,0 +1,428 @@
...
+ * 
+ */
+@Service
+public class ConceptManager implements IConceptManager {
+
+    @Autowired
+    private IConceptDBManager client;
+
+    @Autowired
+    private ILuceneUtility luceneUtility;
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java]: I'd use the search manager instead of connecting directly with the lucene utility and let the searchmanager manage the connection with lucene. This way, if we decide to switch to Solr or Elasticsearch or so later on, just one class has to be changed, not all of them. Or, thinking of it, you could also rename this class to Indexsoemthing (e.g. IndexUtility, or IndexService), so that this class could simply be replaced by a different implementation if we decide to change the indexer.

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/SearchManager.java [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import edu.asu.conceptpower.servlet.core.ConceptEntry;
+import edu.asu.conceptpower.servlet.core.ISearchManager;
+import edu.asu.conceptpower.servlet.exceptions.LuceneException;
+import edu.asu.conceptpower.servlet.lucene.impl.LuceneUtility;
+
+@Service
+public class SearchManager implements ISearchManager {
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/SearchManager.java]: javadoc

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+            createDocuments(iterator, dict, writer);
+
+            iterator = dict.getIndexWordIterator(POS.ADJECTIVE);
+            createDocuments(iterator, dict, writer);
+
+            iterator = dict.getIndexWordIterator(POS.VERB);
+            createDocuments(iterator, dict, writer);
+
+            writer.commit();
+
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: what about all the concepts that are not in Wordnet (but in the Db4o database)?

**Comment by jdamerow** [comment]
I think I look at the major classes you've changed, or did I miss something?

Regarding the front end, currently it looks like this:
image

It should be changed as follows:
- Page heading should "Conceptpower Index"
- Tab title should be "Index"

The page should show the following info:
- last time the indexer was run
- number of indexed words
- keep a history of both values so that we could add a graphic showing the development of indexed words

Also:
- when starting the indexer, the website should have a little spinner that's running until the indexer is done (do an ajax call to find out when the indexer is done).
- while the indexer is running it should not be possible to start it a second time. the LuceneUtility class needs to be thread-safe.
- when the indexer is started although there already exists an index, the index should simply be updated

**Comment by jdamerow** [comment]
Otherwise, good work :)

File: Conceptpower+Spring/pom.xml (hunk 1 above)
Comment (mkarthik90) [inline Conceptpower+Spring/pom.xml]: @jdamerow  I was working on windows machine sometime back. Changed back to the correct location

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java (hunk 15 above)
Comment (mkarthik90) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/core/impl/ConceptManager.java]: @jdamerow This is because in case of spaces in words its not able to search. I tried with white space analyzer as well as standard analyzer

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java (hunk 7 above)
Comment (mkarthik90) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: @jdamerow  I tried this method but I am getting lock issues

Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: did you close the other writer first?

Comment (mkarthik90) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: Yes I closed the writer and assigned it with new writer. But I got an ObjectAlreadyClosed Exception when I tried to access the writer in next processing.

Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: hm, works for me just fine. Did you recreate the searcher as well? Otherwise, it'll try to use a closed reader.
But looking at it again, just use 
```
writer.deleteAll()
```

**Comment by mkarthik90** [comment]
@jdamerow  Fixed the review comments. Please let me know if the changes are fine

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 21]
```diff
@@ -0,0 +1,463 @@
...
+                Object contentOfField = field.get(entry);
+                if (contentOfField != null) {
+
+                    if (searchFieldAnnotation.isIndexable()) {
+                        doc.add(new StringField(searchFieldAnnotation.lucenefieldName(),
+                                String.valueOf(contentOfField).replace(" ", ""), Field.Store.YES));
+                    } else {
+                        doc.add(new StoredField(searchFieldAnnotation.lucenefieldName(),
+                                String.valueOf(contentOfField).replace(" ", "")));
+                    }
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: why do you replace " " with ""?

**Comment by mkarthik90** [comment]
@jdamerow I fixed the review comments. Could you please check the changes

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 22]
```diff
@@ -0,0 +1,442 @@
...
+    private DatabaseClient databaseClient;
+
+    @Autowired
+    @Qualifier("luceneDAO")
+    private ILuceneDAO luceneDAO;
+
+    @Value("${lucenePath}")
+    private String lucenePath;
+
+    @Value("${hits}")
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: this property should have a more telling name, eg. lucene.numberOfResults

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java [hunk 23]
```diff
@@ -0,0 +1,442 @@
...
+            }
+        }
+
+        try {
+            writer.addDocument(doc);
+            writer.commit();
+        } catch (IOException ex) {
+            throw new LuceneException("Cannot insert concept in lucene. Please retry", ex);
+        }
+        luceneDAO.storeValues(1, LuceneAction.INSERT);
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneUtility.java]: beautiful, so much more compact :)

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneDAO.java [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+
+    ObjectContainer luceneClient;
+
+    private String dbPath;
+
+    @PostConstruct
+    public void init() {
+
+        EmbeddedConfiguration config = Db4oEmbedded.newConfiguration();
+        config.common().objectClass(Timestamp.class).translate(new TSerializable());
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneDAO.java]: do you still need this?

File: Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneDAO.java [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+        luceneClient = Db4oEmbedded.openFile(config, getDbPath());
+    }
+
+    /**
+     * Stores the number of indexed word count along with the timestamp
+     */
+    public void storeValues(long numberOfIndexedWords,String action) {
+        long now = System.currentTimeMillis();
+        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+        IndexingEvent bean = new IndexingEvent(new Timestamp(now), numberOfIndexedWords,action,auth.getName());
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/java/edu/asu/conceptpower/servlet/lucene/impl/LuceneDAO.java]: use just ```new Date()``` here?

**Comment by jdamerow** [comment]
- Index page still needs a heading.
- when one user runs the indexer, it shouldn't be possible to start another index run.

File: Conceptpower+Spring/src/main/webapp/WEB-INF/views/auth/lucene/luceneindex.jsp [hunk 26]
```diff
@@ -0,0 +1,210 @@
...
+<%@ taglib prefix="sec"
+	uri="http://www.springframework.org/security/tags"%>
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core"%>
+<%@ taglib prefix="form" uri="http://www.springframework.org/tags/form"%>
+<head>
+<meta charset="utf-8">
+<meta name="viewport"
+	content="width=device-width, initial-scale=1, maximum-scale=1">
+<meta name="apple-mobile-web-app-capable" content="yes">
+<title>waitMe</title>
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/webapp/WEB-INF/views/auth/lucene/luceneindex.jsp]: title?

File: Conceptpower+Spring/src/main/webapp/WEB-INF/views/auth/lucene/luceneindex.jsp [hunk 27]
```diff
@@ -0,0 +1,210 @@
...
+	border-color: #537FA9 !important;
+	color: #ACD3E8;
+	box-shadow: none !important
+}
+
+.btn.btn-left {
+	float: left;
+	margin: 0 5px 0 0 !important
+}
+</style>
```
Comment (jdamerow) [inline Conceptpower+Spring/src/main/webapp/WEB-INF/views/auth/lucene/luceneindex.jsp]: use the old style for the buttons here, otherwise they won't use the new layout once it's merged.

**Comment by jdamerow** [comment]
I'll close this PR, please reopen it when you're done.
